dojo/problem/redis.py

```diff
@@ -0,0 +1,136 @@
+            "script_ids": list(self.script_ids),
+            "finding_ids": list(self.finding_ids),
+        }
+
+    @staticmethod
+    def from_dict(data):
+        return Problem(
+            name=data["name"],
+            problem_id=data["problem_id"],
+            severity=data["severity"],
+            script_ids=set(data.get("script_ids", [])),
+            finding_ids=set(data.get("finding_ids", [])),
+        )
+
+    @classmethod
+    def load_from_id(cls, problem_id, redis_client):
+        problem_data = redis_client.hget("problems", problem_id)
+        if problem_data:
+            return cls.from_dict(json.loads(problem_data))
+        return None
+
+    def persist(self, redis_client):
+        if not self.finding_ids:
+            redis_client.hdel("problems", self.problem_id)
+        else:
+            redis_client.hset("problems", self.problem_id, json.dumps(self.to_dict()))
+
+    @classmethod
+    def add_finding(cls, finding):
+        cls.remove_finding(int(finding.id))
+        if finding.vuln_id_from_tool and finding.severity != "Info":
+            redis_client = get_redis_client()
+            problem_id = problem_id_b64encode(finding.vuln_id_from_tool)
+            problem = cls.load_from_id(problem_id, redis_client)
+            if not problem:
+                problem = cls(problem_id)
+            if finding.id not in problem.finding_ids:
+                if SEVERITY_ORDER[finding.severity] > SEVERITY_ORDER[problem.severity]:
+                    problem.name = finding.title
+                    problem.severity = finding.severity
+                problem.finding_ids.add(finding.id)
+                problem.script_ids.add(finding.vuln_id_from_tool)
+            problem.persist(redis_client)
+            redis_client.hset("id_to_problem", finding.id, problem_id)
+
+    @classmethod
+    def remove_finding(cls, finding_id):
+        redis_client = get_redis_client()
+        if redis_client.exists("problems") and redis_client.exists("id_to_problem"):
+            problem_id = redis_client.hget("id_to_problem", finding_id)
+            if problem_id:
+                problem = cls.load_from_id(problem_id, redis_client)
+                if problem:
+                    # This is why we do not use Redis set functionality to store Findings associated with a Problem.
+                    # We need to iterate over Findings anyway whenever a Finding is deleted or changed to update the Problem definition.
+                    problem.finding_ids.remove(finding_id)
+                    findings = Finding.objects.filter(id__in=problem.finding_ids)
+                    severity, name, script_ids = "Info", "", set()
+                    for finding in findings:
+                        script_ids.add(finding.vuln_id_from_tool)
+                        if SEVERITY_ORDER[finding.severity] > SEVERITY_ORDER[severity]:
+                            name = finding.title
+                            severity = finding.severity
```

Comment:
We should refactor in such a way that we only call `Problem.update_name_sev()` when the *current* "main" `Finding` is removed. If a `Finding` that is not the worst is removed, we can just keep as-is. This requires that `finding.severity` is fixed... but if it is not fixed, then other parts of the code would need to change too and we would need to update the "main" `Finding` every time we load the `Problem`.